File: Java/combination-sum-ii.java
//  Combination Sum II

// Time O(N!), Space O(N)
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<List<Integer>>();
        List<Integer> comb = new ArrayList<Integer>();
        Arrays.sort(candidates);
        dfs(candidates, target, 0, comb, ans);
        return ans;
    }
    
    private void dfs(int[] candidates, int target, int start, List<Integer> comb, List<List<Integer>> ans) {
        if (target < 0)
            return;
        if (target == 0)
            ans.add(new ArrayList(comb));
        int prev = -1;
        for (int i=start; i<candidates.length; i++) {
            if (candidates[i] == prev)
                continue;
            prev = candidates[i];
            comb.add(candidates[i]);
            dfs(candidates, target-candidates[i], i+1, comb, ans);
            comb.remove(comb.size()-1);
        }
    }
}

File: leetcode/combination-sum-ii.cc
//  Combination Sum II

class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> comb;
        sort(candidates.begin(), candidates.end());
        dfs(candidates, target, 0, comb, res);
        return res;
    }
    
    void dfs(vector<int> &candidates, int target, int cur, 
                vector<int> &comb, vector<vector<int>> &res) {
        if (target == 0) {
            if (find(res.begin(), res.end(), comb) == res.end())
                res.push_back(comb);
            return;
        }
        for (int i=cur; i<candidates.size(); i++) {
            if (candidates[i] > target) break;
            comb.push_back(candidates[i]);
            dfs(candidates, target-candidates[i], i+1, comb, res);
            comb.pop_back();
        }
    }
};